### PR TITLE
ddl: loose alter column with foreign key in some cases

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/22820

### What is changed and how it works?

What's Changed:
when changing a column with FK, if its field type hasn't been changed, it can be allowed to run.

For MySQL:
For varchar type, a column with FK can also be allowed to expand the flen size.
But for decimal type, a column with FK can't be allowed to modify the flen&decimal in any case.

**It needs more detailed control here, in this PR, we just banned all the changes of field type in a column with FK.**

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- ddl: loose alter column with fk in some cases.

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -3765,6 +3761,15 @@ func (d *ddl) getModifiableColumnJob(ctx sessionctx.Context, ident ast.Ident, or
 		return nil, errors.Trace(err)
 	}
 
+	// Check the column with foreign key, waiting for the default flen and decimal.
+	if fkInfo := getColumnForeignKeyInfo(originalColName.L, t.Meta().ForeignKeys); fkInfo != nil {
+		// For now we strongly ban the all column type change for column with foreign key.
+		// Actually mysql support change column with foreign key from varchar(m) -> varchar(m+t) and t > 0.
```
Comment (zimulala) [inline ddl/ddl_api.go:3767]: s/mysql/MySQL

File: ddl/column_type_change_test.go [hunk 2]
```diff
@@ -1684,3 +1684,40 @@ func (s *testColumnTypeChangeSuite) TestChangingColOriginDefaultValue(c *C) {
...
+	tk.MustExec("alter table orders modify user_id int default -1 comment \"haha\"")
+	tbl = testGetTableByName(c, tk.Se, "test", "orders")
+	c.Assert(tbl.Meta().Columns[1].Comment, Equals, "haha")
+	c.Assert(tbl.Meta().Columns[1].DefaultValue.(string), Equals, "-1")
+
+	prepare()
+	tk.MustGetErrCode("alter table orders modify user_id bigint", mysql.ErrFKIncompatibleColumns)
+
+	tk.MustExec("drop table if exists orders")
+	tk.MustExec("drop table if exists users")
```
Comment (zimulala) [inline ddl/column_type_change_test.go:1722]:
```suggestion
	tk.MustExec("drop table if exists orders, users")
```

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -3765,6 +3761,15 @@ func (d *ddl) getModifiableColumnJob(ctx sessionctx.Context, ident ast.Ident, or
 		return nil, errors.Trace(err)
 	}
 
+	// Check the column with foreign key, waiting for the default flen and decimal.
+	if fkInfo := getColumnForeignKeyInfo(originalColName.L, t.Meta().ForeignKeys); fkInfo != nil {
```
Comment (zimulala) [inline ddl/ddl_api.go:3765]: Why we change the order of this code? Do we have a test to cover this update？

**Review by tangenta — COMMENTED** [review]
/lgtm

File: ddl/ddl_api.go (hunk 3 above)
Comment (AilinKid) [reply ddl/ddl_api.go:3765]: The newCol.Flen was Unspecified until it is correctly assigned in `setCharsetCollationFlenDecimal` lined in 3760

**Review by zimulala — COMMENTED** [review]
/LGTM

**Review by zimulala — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**